File: src/main/java/me/kuku/yuq/logic/impl/QQAILogicImpl.java
package me.kuku.yuq.logic.impl;

import com.alibaba.fastjson.JSONArray;
import com.alibaba.fastjson.JSONObject;
import me.kuku.yuq.entity.ConfigEntity;
import me.kuku.yuq.logic.AILogic;
import me.kuku.yuq.pojo.Result;
import me.kuku.yuq.service.ConfigService;
import me.kuku.yuq.utils.BotUtils;
import me.kuku.yuq.utils.MD5Utils;
import me.kuku.yuq.utils.OkHttpUtils;
import okhttp3.FormBody;

import javax.inject.Inject;
import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.util.*;

@SuppressWarnings("unused")
public class QQAILogicImpl implements AILogic {
    @Inject
    private ConfigService configService;

    private String getSign(Map<String, String> map, String appKey){
        TreeMap<String, String> treeMap = new TreeMap<>(map);
        StringBuilder sb = new StringBuilder();
        for (Map.Entry<String, String> entry: treeMap.entrySet()){
            try {
                sb.append(String.format("%s=%s&", entry.getKey(), URLEncoder.encode(entry.getValue(), "utf-8")));
            } catch (UnsupportedEncodingException e) {
                e.printStackTrace();
            }
        }
        sb.append("app_key=").append(appKey);
        return MD5Utils.toMD5(sb.toString()).toUpperCase();
    }

    private FormBody addParams(Map<String, String> otherParams){
        ConfigEntity configEntity1 = configService.findByType("qqAIAppId");
        ConfigEntity configEntity2 = configService.findByType("qqAIAppKey");
        String appId;
        String appKey;
        if (configEntity1 == null || configEntity2 == null) {
            appId = "";
            appKey = "";
        }else{
            appId = configEntity1.getContent();
            appKey = configEntity2.getContent();
        }
        Map<String, String> map = new HashMap<>();
        map.put("app_id", appId);
        map.put("time_stamp", String.valueOf(new Date().getTime() / 1000));
        map.put("nonce_str", BotUtils.randomStr(16));
        FormBody.Builder builder = new FormBody.Builder();
        map.putAll(otherParams);
        String sign = getSign(map, appKey);
        map.put("sign", sign);
        for (Map.Entry<String, String> entry: map.entrySet()){
            builder.add(entry.getKey(), entry.getValue());
        }
        return builder.build();
    }

    private String urlToBase64(String imageUrl) throws IOException {
        byte[] bytes = OkHttpUtils.getBytes(imageUrl);
        return Base64.getEncoder().encodeToString(bytes);
    }

    @Override
    public boolean pornIdentification(String imageUrl) throws IOException {
        String baseStr = urlToBase64(imageUrl);
        Map<String, String> map = new HashMap<>();
        map.put("image", baseStr);
        JSONObject jsonObject = OkHttpUtils.postJson("https://api.ai.qq.com/fcgi-bin/vision/vision_porn",
                addParams(map));
        if (jsonObject.getInteger("ret").equals(0)){
            JSONArray jsonArray = jsonObject.getJSONObject("data").getJSONArray("tag_list");
//            Integer normal = jsonArray.getJSONObject(0).getInteger("tag_confidence");
//            Integer hot = jsonArray.getJSONObject(1).getInteger("tag_confidence");
            Integer porn = jsonArray.getJSONObject(2).getInteger("tag_confidence");
            return porn > 83 /*|| hot > normal*/;
        }else return false;
    }

    @Override
    public String generalOCR(String imageUrl) throws IOException {
        String baseStr = urlToBase64(imageUrl);
        Map<String, String> map = new HashMap<>();
        map.put("image", baseStr);
        JSONObject jsonObject = OkHttpUtils.postJson("https://api.ai.qq.com/fcgi-bin/ocr/ocr_generalocr", addParams(map));
        if (jsonObject.getInteger("ret").equals(0)){
            JSONArray jsonArray = jsonObject.getJSONObject("data").getJSONArray("item_list");
            if (jsonArray.size() == 0) return "啥文字也没有识别到";
            StringBuilder sb = new StringBuilder();
            for (int i = 0 ; i < jsonArray.size(); i++){
                JSONObject singleJsonObject = jsonArray.getJSONObject(i);
                sb.append(singleJsonObject.getString("itemstring")).append("\n");
            }
            return sb.deleteCharAt(sb.length() - 1).toString();
        }else return jsonObject.getString("msg");
    }

    @Override
    public String textChat(String question, String session) throws IOException {
        Map<String, String> map = new HashMap<>();
        map.put("session", session);
        map.put("question", question);
        JSONObject jsonObject = OkHttpUtils.postJson("https://api.ai.qq.com/fcgi-bin/nlp/nlp_textchat",
                addParams(map));
        Integer ret = jsonObject.getInteger("ret");
        switch (ret){
            case 0: return jsonObject.getJSONObject("data").getString("answer");
            case 16385: return "没有填入appid， 请联系机器人主人填写";
            case 16394: return "没有发现匹配的答案";
            default: jsonObject.getString("msg");
        }
        return null;
    }

    @Override
    public Result<byte[]> voiceSynthesis(String text) throws IOException {
        Map<String, String> map = new HashMap<>();
        map.put("speaker", "5");
        map.put("format", "2");
        map.put("volume", "0");
        map.put("speed", "100");
        map.put("text", text);
        map.put("aht", "0");
        map.put("apc", "58");
        JSONObject jsonObject = OkHttpUtils.postJson("https://api.ai.qq.com/fcgi-bin/aai/aai_tts", addParams(map));
        if (jsonObject.getInteger("ret") == 0){
            String base64 = jsonObject.getJSONObject("data").getString("speech");
            return Result.success(Base64.getDecoder().decode(base64));
        }else return Result.failure(jsonObject.getString("msg"), null);
    }

    @Override
    public String voiceIdentify(String voiceUrl) throws IOException {
        return null;
    }
}

File: src/main/kotlin/me/kuku/yuq/logic/impl/QQAILogicImpl.kt
package me.kuku.yuq.logic.impl

import com.alibaba.fastjson.JSONObject
import me.kuku.yuq.logic.QQAILogic
import me.kuku.yuq.pojo.CommonResult
import me.kuku.yuq.service.ConfigService
import me.kuku.yuq.utils.BotUtils
import me.kuku.yuq.utils.MD5Utils
import me.kuku.yuq.utils.OkHttpClientUtils
import me.kuku.yuq.utils.removeSuffixLine
import okhttp3.FormBody
import java.net.URLEncoder
import java.util.*
import javax.inject.Inject

class QQAILogicImpl: QQAILogic {

    @Inject
    private lateinit var configService: ConfigService

    private fun getSign(map: Map<String, String>, appKey: String): String{
        val treeMap = TreeMap<String, String>()
        treeMap.putAll(map)
        val sb = StringBuilder()
        for ((k, v) in treeMap){
            sb.append("$k=${URLEncoder.encode(v, "utf-8")}&")
        }
        sb.append("app_key=$appKey")
        return MD5Utils.toMD5(sb.toString()).toUpperCase()
    }

    private fun addParams(otherParams: Map<String, String>): FormBody {
        val configEntity1 = configService.findByType("qqAIAppId")
        val configEntity2 = configService.findByType("qqAIAppKey")
        val appId = configEntity1?.content ?: ""
        val appKey = configEntity2?.content ?: ""
        val map = mutableMapOf("app_id" to appId, "time_stamp" to (Date().time / 1000).toString(), "nonce_str" to BotUtils.randomStr(16))
        val builder = FormBody.Builder()
        map.putAll(otherParams)
        val sign = this.getSign(map, appKey)
        map["sign"] = sign
        for ((k , v) in map){
            builder.add(k, v)
        }
        return builder.build()
    }

    private fun urlToBase64(imageUrl: String): String{
        val response = OkHttpClientUtils.get(imageUrl)
        val bytes = OkHttpClientUtils.getBytes(response)
        return Base64.getEncoder().encodeToString(bytes)
    }

    override fun pornIdentification(imageUrl: String): Boolean{
        val baseStr = this.urlToBase64(imageUrl)
        val response = OkHttpClientUtils.post("https://api.ai.qq.com/fcgi-bin/vision/vision_porn",
                addParams(mapOf("image" to baseStr)))
        val jsonObject = OkHttpClientUtils.getJson(response)
        return if (jsonObject.getInteger("ret") == 0){
            val jsonArray = jsonObject.getJSONObject("data").getJSONArray("tag_list")
            /*val normal = jsonArray.getJSONObject(0).getInteger("tag_confidence")
            val hot = jsonArray.getJSONObject(1).getInteger("tag_confidence")*/
            val porn = jsonArray.getJSONObject(2).getInteger("tag_confidence")
            porn > 83/* || hot > normal*/
        }else false
    }

    override fun generalOCR(imageUrl: String): String{
        val baseStr = this.urlToBase64(imageUrl)
        val response = OkHttpClientUtils.post("https://api.ai.qq.com/fcgi-bin/ocr/ocr_generalocr",
                addParams(mapOf("image" to baseStr)))
        val jsonObject = OkHttpClientUtils.getJson(response)
        return if (jsonObject.getInteger("ret") == 0){
            val jsonArray = jsonObject.getJSONObject("data").getJSONArray("item_list")
            if (jsonArray.isEmpty()) return "啥文字也没有识别到！！"
            val sb = StringBuilder()
            jsonArray.forEach {
                val singleJsonObject = it as JSONObject
                sb.appendLine(singleJsonObject.getString("itemstring"))
            }
            sb.removeSuffixLine().toString()
        }else jsonObject.getString("msg")
    }

    override fun textChat(question: String, session: String): String{
        val response = OkHttpClientUtils.post("https://api.ai.qq.com/fcgi-bin/nlp/nlp_textchat",
                addParams(mapOf("session" to session, "question" to question)))
        val jsonObject = OkHttpClientUtils.getJson(response)
        return when (jsonObject.getInteger("ret")){
            0 -> jsonObject.getJSONObject("data").getString("answer")
            16385 -> "您还没有填入appid！！"
            16394 -> "没有发现匹配的答案！！"
            else -> jsonObject.getString("msg")
        }
    }

    override fun echoSpeechRecognition(url: String): String {
        val speechResponse = OkHttpClientUtils.get(url)
        val bytes = OkHttpClientUtils.getBytes(speechResponse)
        val b64Str = Base64.getEncoder().encodeToString(bytes)
        val response = OkHttpClientUtils.post(
                "https://api.ai.qq.com/fcgi-bin/aai/aai_asr",
                addParams(
                        mapOf(
                                "format" to "3",
                                "speech" to b64Str,
                                "rate" to "16000"
                        )
                )
        )
        val jsonObject = OkHttpClientUtils.getJson(response)
        println(jsonObject)
        return ""
    }

    override fun aiLabSpeechRecognition(url: String): String {
        val speechResponse = OkHttpClientUtils.get(url)
        val bytes = OkHttpClientUtils.getBytes(speechResponse)
        val b64Str = Base64.getEncoder().encodeToString(bytes)
        val response = OkHttpClientUtils.post(
                "https://api.ai.qq.com/fcgi-bin/aai/aai_asrs",
                addParams(
                        mapOf(
                                "format" to "3",
                                "rate" to "16000",
                                "seq" to "0",
                                "len" to bytes.size.toString(),
                                "end" to "1",
                                "speech_id" to "123213123",
                                "speech_chunk" to b64Str
                        )
                )
        )
        val jsonObject = OkHttpClientUtils.getJson(response)
        println(jsonObject)
        return ""
    }

    override fun voiceSynthesis(text: String): CommonResult<ByteArray> {
        val response = OkHttpClientUtils.post("https://api.ai.qq.com/fcgi-bin/aai/aai_tts", addParams(mapOf(
                "speaker" to "5",
                "format" to "2",
                "volume" to "0",
                "speed" to "100",
                "text" to text,
                "aht" to "0",
                "apc" to "58"
        )))
        val jsonObject = OkHttpClientUtils.getJson(response)
        return if (jsonObject.getInteger("ret") == 0) {
            val base64 = jsonObject.getJSONObject("data").getString("speech")
            CommonResult(200, "", Base64.getDecoder().decode(base64))
        }else CommonResult(500, jsonObject.getString("msg"))
    }
}